Moving sample files to cpp sdk repo (#5947)
Azure/azure-sdk-for-cpp: samples/integration/cmake-vcpkg/README.md

# Develop an Azure C++ App using vcpkg and CMake

## Perquisite

1. Any Text Editor
2. A terminal
3. A C++ compiler
    - Windows: [MSVC](https://learn.microsoft.com/cpp/build/vscpp-step-0-installation?view=msvc-170)
4. [git](https://git-scm.com/downloads)

## Steps

### Install CMake

> - Go to the [CMake Downloads page](https://cmake.org/download/) where you can either download the latest version of the CMake source or an installer for your OS. I'd recommend using the installer for simplicity. When asked select the option to add CMake to your PATH.

Reviewer: ```suggestion
- Go to the [CMake Downloads page](https://cmake.org/download/) to download an installer for your OS. The installer approach is recommended for simplicity. Optionally you can download the latest version of the CMake source instead. When asked, select the option to add CMake to your PATH.
```

> - Once you've downloaded and run through the installer program run the following command to verify CMake has been installed properly:

Reviewer: ```suggestion
- After installing CMake, run the following command to verify CMake has been installed properly:
```

```bash
cmake --version
```

- You should get an output denoting the currently install version of CMake. Like this:

```bash
cmake version 3.30.2
```

- If you don't see the expected output you may need to restart your computer.

### Install vcpkg

> - To install vcpkg you'll first need to clone the vcpkg repo. I recommend cloning vcpkg to a central location on your development environment and not in your C++ Project directory. In this example I'll clone vcpkg to my home dir.

Reviewer: ```suggestion
- To install vcpkg you'll first need to clone the vcpkg repo. The recommended approach is to clone vcpkg to a central location on your development environment and not in your C++ Project directory. In this example vcpkg is cloned to the home dir.
```

```bash

> cd ~

Reviewer 2: You should be able to use / or \ in the environment variable. 
Windows accepts either and cmake gets annoyed with \.

git clone https://github.com/microsoft/vcpkg.git
```

- Once the vcpkg repo is cloned traverse into the new directory and run the bootstrap-vcpkg script.
  - If you're on Windows run:

  ```bash
  cd .\vcpkg\
  .\bootstrap-vcpkg.bat
  ```

  - If you're on MacOS or Linux, run:
  
  ```bash
  cd vcpkg
  ./bootstrap-vcpkg.sh
  ```

- After bootstrapping vcpkg, add it to your path so you can access the vcpkg executable from your project directory. Remember to replace the path in the command example with the path to the vcpkg directory you cloned earlier.
  - On Windows Powershell, enter:
  
  ```bash
  $env:Path = "$env:Path;C:\path\to\vcpkg"
  ```

  - On MacOS or Linux systems, enter:
  
  ```bash
  PATH=$PATH:/path/to/vcpkg
  ```

- Traverse back to your project directory and run the following to verify the vcpkg directory was added to your path:

```bash
vcpkg --version
```

> - You should get an output resembling the following (The hash at the end of the line does not need to match exactly):

Reviewer 2: The version will also almost certainly not match as well - this version works today but won't in a couple of months.

cmake -B ./build
```

- You should see the build directory populate with some directories and files.
- Now lets try building the project with:

```bash
cmake --build ./build
```

- The output of the build should contain a line stating where the executable that was built was placed. By default it should be in your `build` directory under a new `Debug` directory with the name `azure_sample.exe` if your on Windows or `azure_sample` if your on MacOS or Linux.
- Lets try running our new executable.
  - If your on Windows Powershell, enter:
  
  ```pwsh
  .\build\Debug\azure_sample.exe
  ```

  - If your on MacOS or Linux, enter:
  
  ```bash
  ./build/azure_sample
  ```

- You should get the following out put:

```bash
Hello World!
```

### Install packages with vcpkg

- In this example we'll be using vcpkg in what's known as manifest mode.
- First we'll start by moving back to the root directory of our project, and creating a new vcpkg application with the following command:

```bash
vcpkg new --application
```

- You should now have a `vcpkg.json` and `vcpkg-configuration.json` file in your project directory with the following contents:
  - `vcpkg.json`
  
  ```json
  {}
  ```

>   - `vcpkg-configuration.json` *It's okay if the hash on the `baseline` property doesn't match exactly.*

Reviewer 2: I wouldn't even add "exactly" - it's almost certainly never going to match and it might match radically differently.

        "name": "microsoft"
      }
    ]
  }
  ```

- Now we can add the Azure Key Vault and Identity libraries from the Azure SDK for C++ to our project, by entering the following command:

```bash
vcpkg add port azure-identity-cpp azure-security-keyvault-secrets-cpp
```

- Your `vcpkg.json` file should now look like this:

```json
{
  "dependencies": [
    "azure-identity-cpp",
    "azure-security-keyvault-secrets-cpp"
  ]
}
```

### Integrate CMake with vcpkg

- To integrate CMake with vcpkg we need to add the `vcpkg.cmake` module to our CMake toolchain. You can do this either by setting `CMAKE_TOOLCHAIN_FILE` in your `CMakeLists.txt` file like this:

```cmake
cmake_minimum_required(VERSION 3.30.0)

# Remember to replace the path below with the path where you cloned vcpkg
set(CMAKE_TOOLCHAIN_FILE "/path/to/vcpkg-root/scripts/buildsystems/vcpkg.cmake")

project(azure_sample VERSION 0.1.0 LANGUAGES C CXX)

add_executable(azure_sample main.cpp)
```

- Or by specifying the toolchain file in the CMake configuration command like this:

```bash
cmake -DCMAKE_TOOLCHAIN_FILE=/path/to/vcpkg-root/scripts/buildsystems/vcpkg.cmake -B build
```

> - I'd recommend specifying the toolchain file in the `CMakeLists.txt` for now as it will shorten the commands you have to enter.

Reviewer 2: Do you want to discuss cmakepresets or cmakeuserpresets here? That is also a great place to stick the toolchain file definition.
